File: Project-source/forge-verifier/csharp/src/OnlineVerifier.cs
// Online / hybrid composition — parses + verifies signature/expiry, then dispatches
// CRL refresh + heartbeat based on the license's own `mode` field.
//
// Fail-closed policy mirrors Go SDK (forge-verifier/go/online.go):
//   - offline → never touches network
//   - hybrid  → signature MUST pass; CRL + heartbeat best-effort; cached revoke still wins
//   - online  → all three MUST pass; network failure → reject

namespace YourCo.Forge.Verifier;

public sealed class OnlineVerifier
{
    private readonly byte[] _publicKey;
    private readonly HeartbeatClient? _heartbeat;
    private readonly CrlClient? _crl;

    public OnlineVerifier(byte[] publicKey, HeartbeatClient? heartbeat = null, CrlClient? crl = null)
    {
        _publicKey = publicKey;
        _heartbeat = heartbeat;
        _crl = crl;
    }

    public async Task<ForgeVerifier.Result> VerifyAsync(
        string forgePath,
        string fingerprint,
        DateTimeOffset now,
        CancellationToken ct = default)
    {
        var parsed = ForgeVerifier.Parse(forgePath);
        var res = parsed.Verify(_publicKey, now);
        var mode = string.IsNullOrEmpty(parsed.Payload.Mode) ? "offline" : parsed.Payload.Mode;

        switch (mode)
        {
            case "offline":
                return res;

            case "hybrid":
                if (_crl is not null)
                {
                    try { await _crl.RefreshAsync(ct).ConfigureAwait(false); }
                    catch { /* hybrid: ignore network errors; fall back to cached snapshot */ }
                    if (_crl.IsRevoked(parsed.Payload.LicenseId))
                    {
                        throw new ForgeVerifier.ForgeException("license revoked (CRL)",
                            new ForgeVerifier.Result
                            {
                                Status = ForgeVerifier.Status.Revoked,
                                LicenseId = parsed.Payload.LicenseId,
                                ExpiresAt = parsed.Payload.ExpiresAt,
                                Binding = parsed.Payload.Binding,
                                FingerprintMustMatch = parsed.Payload.BoundFingerprint,
                            });
                    }
                }
                if (_heartbeat is not null)
                {
                    try { await _heartbeat.SendAsync(parsed.Payload.LicenseId, fingerprint, ct).ConfigureAwait(false); }
                    catch { /* best-effort */ }
                }
                return res;

            case "online":
                if (_crl is null || _heartbeat is null)
                {
                    throw new ForgeVerifier.ForgeException(
                        "online mode requires CrlClient + HeartbeatClient",
                        ForgeVerifier.Status.Malformed);
                }
                try { await _crl.RefreshAsync(ct).ConfigureAwait(false); }
                catch (Exception e)
                {
                    throw new ForgeVerifier.ForgeException(
                        $"online: CRL refresh failed: {e.Message}",
                        ForgeVerifier.Status.SignatureInvalid);
                }
                if (_crl.IsRevoked(parsed.Payload.LicenseId))
                {
                    throw new ForgeVerifier.ForgeException("license revoked",
                        new ForgeVerifier.Result
                        {
                            Status = ForgeVerifier.Status.Revoked,
                            LicenseId = parsed.Payload.LicenseId,
                            ExpiresAt = parsed.Payload.ExpiresAt,
                            Binding = parsed.Payload.Binding,
                            FingerprintMustMatch = parsed.Payload.BoundFingerprint,
                        });
                }
                try
                {
                    var hb = await _heartbeat.SendAsync(parsed.Payload.LicenseId, fingerprint, ct)
                        .ConfigureAwait(false);
                    if (!string.IsNullOrEmpty(hb.LicenseStatus) && hb.LicenseStatus != "active")
                    {
                        throw new ForgeVerifier.ForgeException(
                            $"online: server reports license_status={hb.LicenseStatus}",
                            ForgeVerifier.Status.SignatureInvalid);
                    }
                }
                catch (ForgeVerifier.ForgeException) { throw; }
                catch (Exception e)
                {
                    throw new ForgeVerifier.ForgeException(
                        $"online: heartbeat failed: {e.Message}",
                        ForgeVerifier.Status.SignatureInvalid);
                }
                return res;

            default:
                throw new ForgeVerifier.ForgeException(
                    $"unknown license mode: {mode}",
                    ForgeVerifier.Status.Malformed);
        }
    }
}

File: Project-source/forge-verifier/java/src/main/java/com/yourco/forge/verifier/OnlineVerifier.java
package com.yourco.forge.verifier;

import com.yourco.forge.verifier.ForgeVerifier.ForgeException;
import com.yourco.forge.verifier.ForgeVerifier.ForgeFile;
import com.yourco.forge.verifier.ForgeVerifier.Result;
import com.yourco.forge.verifier.ForgeVerifier.Status;

import java.nio.file.Path;
import java.time.Instant;

/**
 * Composes {@link ForgeVerifier#verify}, {@link CrlClient} and {@link HeartbeatClient}
 * into one entry-point. The license's own {@code mode} field selects which network
 * calls run; callers don't need to know.
 *
 * <p>Fail-closed policy:
 * <ul>
 *   <li>{@code offline} — only signature + expiry; never touches the network.</li>
 *   <li>{@code hybrid} — signature + expiry MUST pass; CRL + heartbeat best-effort.
 *       Network failure does not invalidate; an entry in the cached CRL does.</li>
 *   <li>{@code online} — signature + expiry MUST pass AND heartbeat MUST succeed AND
 *       license MUST NOT be in CRL. Network failure → reject.</li>
 * </ul>
 */
public final class OnlineVerifier {

    private final byte[] publicKey;
    private final HeartbeatClient heartbeat;
    private final CrlClient crl;

    public OnlineVerifier(byte[] publicKey, HeartbeatClient heartbeat, CrlClient crl) {
        this.publicKey = publicKey;
        this.heartbeat = heartbeat;
        this.crl = crl;
    }

    public Result verify(Path forgeFile, String fingerprint, Instant now) throws ForgeException {
        ForgeFile parsed = ForgeVerifier.parse(forgeFile);
        Result res = parsed.verify(publicKey, now);
        String mode = parsed.payload.mode == null ? "offline" : parsed.payload.mode;
        switch (mode) {
            case "offline":
            case "":
                return res;

            case "hybrid":
                if (crl != null) {
                    try {
                        crl.refresh();
                    } catch (Exception ignored) {
                        // hybrid is best-effort on network — fall back to cached snapshot
                    }
                    if (crl.isRevoked(parsed.payload.licenseId)) {
                        throw new ForgeException("license revoked (CRL)",
                                new Result.Builder().status(Status.REVOKED)
                                        .licenseId(parsed.payload.licenseId)
                                        .expiresAt(parsed.payload.expiresAt)
                                        .binding(parsed.payload.binding)
                                        .fingerprintMustMatch(parsed.payload.boundFingerprint)
                                        .build());
                    }
                }
                if (heartbeat != null) {
                    try {
                        heartbeat.send(parsed.payload.licenseId, fingerprint);
                    } catch (Exception ignored) {
                        // hybrid heartbeat is best-effort
                    }
                }
                return res;

            case "online":
                if (crl == null || heartbeat == null) {
                    throw new ForgeException("online mode requires CrlClient + HeartbeatClient",
                            Status.MALFORMED);
                }
                try {
                    crl.refresh();
                } catch (Exception e) {
                    throw new ForgeException("online: CRL refresh failed: " + e.getMessage(),
                            Status.SIGNATURE_INVALID);
                }
                if (crl.isRevoked(parsed.payload.licenseId)) {
                    throw new ForgeException("license revoked",
                            new Result.Builder().status(Status.REVOKED)
                                    .licenseId(parsed.payload.licenseId)
                                    .expiresAt(parsed.payload.expiresAt)
                                    .binding(parsed.payload.binding)
                                    .fingerprintMustMatch(parsed.payload.boundFingerprint)
                                    .build());
                }
                try {
                    HeartbeatClient.Response hb =
                            heartbeat.send(parsed.payload.licenseId, fingerprint);
                    if (hb.licenseStatus != null && !hb.licenseStatus.isEmpty()
                            && !"active".equals(hb.licenseStatus)) {
                        throw new ForgeException(
                                "online: server reports license_status=" + hb.licenseStatus,
                                Status.SIGNATURE_INVALID);
                    }
                } catch (ForgeException fe) {
                    throw fe;
                } catch (Exception e) {
                    throw new ForgeException("online: heartbeat failed: " + e.getMessage(),
                            Status.SIGNATURE_INVALID);
                }
                return res;

            default:
                throw new ForgeException("unknown license mode: " + mode, Status.MALFORMED);
        }
    }
}
